### PR TITLE
Implemented support for #EXT-X-MEDIA:TYPE=SUBTITLES tracks (LIVE and VOD)

### PR DESCRIPTION
**Pull request by neilrackett**
Bonjour, et j'espère vous n'avez pas été affecté par l'atrocité récente à Paris.

I have added support for WebVTT subtitles tracks for both live and VOD streams, modelled on the current alternative audio implementation, and resolved a small bug in the `Level` class.

Every effort has been made to ensure that all new events, properties and settings are in keeping with existing code style and naming conventions, and all new classes, properties and methods have basic ASDocs.
## Overview

Subtitles tracks are detected and selected in the same way as alternative audio at present. When a subtitles track is selected, Flashls will dispatch a `SUBTITLES_CHANGE` events to indicate that new subtitles should be displayed.

There is currently no OSMF or Flowplayer specific integration.
## New HLS class properties

_Subtitles_
- `subtitlesPlaylistTracks:Vector.<SubtitlesPlaylistTrack>` subtitles equivalent to `altAudioTracks`
- `subtitlesTracks:Vector.<SubtitlesTrack>`subtitles equivalent to `audioTracks`
- `subtitlesTrack:int` subtitles equivalent to `audioTrack`
- `numSubtitlesTracks:uint` number of available subtitles tracks
- `defaultSubtitlesTrack:int` detected default subtitles track ID (`-1` if none)
- `forcedSubtitlesTrack:int` detected forced subtitles track ID (`-1` if none)
- `hasForcedSubtitles:Boolean` whether or not there are forced subtitles in the current playlist

_Other_
- `numAudioTracks:uint` number of available audio tracks
## New HLSEvent class events and properties

_Events_
- `SUBTITLES_TRACK_SWITCH` when the selected subtitles track changes
- `SUBTITLES_LEVEL_LOADING` when the subtitles level is loading
- `SUBTITLES_LEVEL_LOADED` when the subtitles level is loaded
- `SUBTITLES_CHANGE` when the current subtitles change

_Properties_
- `subtitlesTrack:int` the subtitles track ID related to the current event
- `subtitles:Subtitles` model containing subtitles related to the current event
## New HLSSettings properties
- `autoSelectSubtitles:Boolean` whether or not to select subtitles tracks flagged `AUTOSELECT=YES` that match the current system locale
- `autoSelectDefaultSubtitles:Boolean` whether or not to select subtitles tracks flagged `DEFAULT=YES` (selected in preference to `AUTOSELECT=YES`)
- `autoSelectForcedSubtitles:Boolean` whether or not to select subtitles tracks flagged `FORCED=YES` (selected in preference to any other)
## Code example

``` actionscript3
hls.addEventListener(HLSEvent.SUBTITLES_CHANGE, function(e:HLSEvent):void
{
    trace(e.subtitles.text);
});

if (hls.numSubtitlesTracks) hls.subtitlesTrack = 0;
```
## Demo

[Click here](http://demo.mesmotronic.com/flashls/subtitles) to see a basic demonstration of Flashls with subtitles.
## Bug fix

In the `Level` class, the `getFragmentfromSeqNum` method has been updated to prevent a return value `-2` from `getIndexfromSeqNum(seqnum)` crashing Flash Player, which currently happend regularly with live streams.

### CODE REVIEW

**Comment by mangui** [comment]
really nice feature !  and well explained.
congrats :+1: 
I will review it later this week

File: src/org/mangui/hls/HLS.as [hunk 1]
```diff
@@ -30,6 +36,9 @@ package org.mangui.hls {
         private var _levelLoader : LevelLoader;
         private var _altAudioLevelLoader : AltAudioLevelLoader;
         private var _audioTrackController : AudioTrackController;
+		private var _subtitlesLevelLoader : SubtitlesLevelLoader;
```
Comment (mangui) [inline src/org/mangui/hls/HLS.as]: small indent issue here and there

File: src/org/mangui/hls/constant/HLSLoaderTypes.as [hunk 2]
```diff
@@ -10,6 +10,7 @@
         public static const LEVEL_MAIN : int = 1;
     	// playlist / level loader
         public static const LEVEL_ALTAUDIO : int = 2;
+        public static const LEVEL_SUBTITLES : int = 5;
```
Comment (mangui) [inline src/org/mangui/hls/constant/HLSLoaderTypes.as]: could you order it (after ALT AUDIO) for sake of clarity ?

File: src/org/mangui/hls/event/HLSEvent.as [hunk 3]
```diff
@@ -97,8 +106,12 @@ package org.mangui.hls.event {
         public var streamType: String;
         /** The current audio track **/
         public var audioTrack : int;
+        /** The current subtitles track **/
+        public var subtitlesTrack : int;
```
Comment (mangui) [inline src/org/mangui/hls/event/HLSEvent.as:112]: never used ?

File: src/org/mangui/hls/loader/AltAudioFragmentLoader.as [hunk 4]
```diff
@@ -657,7 +657,8 @@ package org.mangui.hls.loader {
             if (_loadingState == LOADING_IDLE)
                 return;
             var hlsError : HLSError;
-            var fragData : FragmentData = _fragCurrent.data;
+			var fragData : FragmentData = _fragCurrent.data;
```
Comment (mangui) [inline src/org/mangui/hls/loader/AltAudioFragmentLoader.as]: indent

File: src/org/mangui/hls/loader/LevelLoader.as [hunk 5]
```diff
@@ -111,7 +114,7 @@ package org.mangui.hls.loader {
                     }
                     // try next redundant stream
                     if (_levels[_loadLevel].redundantStreamId < _levels[_loadLevel].redundantStreamsNb) {
-                        _levels[_loadLevel].redundantStreamId++;
```
Comment (mangui) [inline src/org/mangui/hls/loader/LevelLoader.as:115]: indent

File: src/org/mangui/hls/event/HLSEvent.as (hunk 3 above)
Comment (neilrackett) [inline src/org/mangui/hls/event/HLSEvent.as:112]: It's used by `SUBTITLES_TRACK_SWITCH` events.

File: src/org/mangui/hls/loader/AltAudioFragmentLoader.as (hunk 4 above)
Comment (neilrackett) [inline src/org/mangui/hls/loader/AltAudioFragmentLoader.as]: Looks like a spaces-vs-tabs issue. I'll wait for you to finish the code review and fix them all together.

File: src/org/mangui/hls/constant/HLSLoaderTypes.as (hunk 2 above)
Comment (neilrackett) [inline src/org/mangui/hls/constant/HLSLoaderTypes.as]: No problem. I wanted to do that originally, but didn't know if changing 3 and 4 would cause problems elsewhere; updated locally, will be in next push.

File: src/org/mangui/hls/loader/LevelLoader.as (hunk 5 above)
Comment (neilrackett) [inline src/org/mangui/hls/loader/LevelLoader.as:115]: I have replaced all tabs with 4 spaces, so that should resolve all of the indentation issues.

**Comment by mangui** [comment]
current logic is using HLSEvent.MEDIA_TIME to look for subtitle matching with current position.
it would be better to handle subtitle as ID3 tag or FRAGMENT_PLAYING event: instead of doing a lookup at each mediaTime  (which could be inaccurate as MEDIA TIME frequency can be configured using HLSSettings as well ...) it would be better and more optimal to inject subtitle as METADATA FLV Tag. thus they would be dispatched at the exact timings through a NetStream event. you can check how FRAGMENT_PLAYING is handled: 

tag is created here : https://github.com/mangui/flashls/blob/24b4eb0668d2671d070246bc32f5238a85239985/src/org/mangui/hls/model/Fragment.as#L60-L88
basically you need to serialize your subtitle content there.

as you can see listener name is set to "onHLSFragmentChange" : https://github.com/mangui/flashls/blob/24b4eb0668d2671d070246bc32f5238a85239985/src/org/mangui/hls/model/Fragment.as#L64

then at the exact timing, NetStream will fire the event, see here : https://github.com/mangui/flashls/blob/24b4eb0668d2671d070246bc32f5238a85239985/src/org/mangui/hls/stream/HLSNetStream.as#L95

**Comment by neilrackett** [comment]
Thanks for the feedback: I'll have a look into the suggested changes to subtitle matching logic later this week or next.

**Comment by neilrackett** [comment]
I started experimenting with your FLV tag suggestion for subtitles timing by implementing a TX3G style `onTextData` NetStream event.

However, although the tags appear to be added to the NetStream, some or all of the events are dispatched immediately, rather than waiting for the appropriate time code, so I'm clearly missing a vital step or detail somewhere.

You can see my updates on the `dev-subtitles-tag` branch here: https://github.com/neilrackett/flashls/commit/ef91c1b66be12e953c6ab8f38d8a5f6440d1cf74

Any hints or help gratefully received!

**Comment by mangui** [comment]
you need to use PTS/DTS (in your case PTS=DTS) instead of startPosition when creating your FLV Tag
for each WebVTT fragment, you should get its start PTS : 

see HLS spec

In order to synchronize timestamps between audio/video and subtitles,
   a X-TIMESTAMP-MAP metadata header SHOULD be added to each WebVTT
   header.  This header maps WebVTT cue timestamps to MPEG-2 (PES)
   timestamps in other Renditions of the Variant Stream.  Its format is:

   X-TIMESTAMP-MAP=LOCAL:<cue time>,MPEGTS:<MPEG-2 time>
   e.g. X-TIMESTAMP-MAP=LOCAL:00:00:00.000,MPEGTS:900000

   The cue timestamp in the LOCAL attribute MAY fall outside the range
   of time covered by the segment.

**Comment by neilrackett** [comment]
Thanks for the info. However, I was concentrating on VOD initially, which invariably uses `X-TIMESTAMP-MAP=MPEGTS:0,LOCAL:00:00:00.000`, so the timestamps would be the same with or without a PTS/DTS offset.

**Comment by mangui** [comment]
well, this is not matching with the spec I am affraid.
using PTS is the only reliable way to deal with metadata/audio/video synchronization

**Comment by neilrackett** [comment]
I think there might be a misunderstanding: even PTS to offset each subtitle, I still see all of the `onTextData` events dispatched at once.

I was wondering if it might be related to the way the tags are appended to `StreamBuffer`? Or whether something needed to be updated to ensure that each of the subtitles are processed at the correct time rather than all together at the start of a sequence? (I don't currently have a complete understanding of how `StreamBuffer` processes the metadata)

**Comment by neilrackett** [comment]
Aha! Looks like I have it about right, just forgot to multiply start timecodes by 1000.

**Comment by mangui** [comment]
close to the optimal solution ? :+1:

**Comment by neilrackett** [comment]
I'm not sure how close, but certainly heading in the right direction!

**Comment by neilrackett** [comment]
Just a quick update for anyone that's interested: so far, the metadata tag implementation (`dev-subs-tag` branch) works for VOD streams playing forward, but there are issues with live and seeking backwards or to a position part way into a subtitle, so there's clearly still work to do.

For now, the media time implementation (`dev` or `dev-subs-time` branch) appears to be working well for both VOD and live streams, so if you need HLS subtitles, that's the version to use.

I'm hoping have time to take another look at `dev-subs-tag` in the not too distant future, but in the mean time, if anyone out there would like to make a contribution to a more optimal solution, all help is gratefully received.

**Comment by neilrackett** [comment]
I'm going close this pull request and reset my `dev` branch to match `mangui/flashls` to enable me to work on other issues (work on subtitles will continue on `dev-subs-time` and `dev-subs-tag` branches).

I'll re-submit the request when a tag based solution is in place.
